src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,300 @@
+ * now minus the cost to spend the current inputs later, plus the amount exceeding the target value. We search the tree to find the
+ * set of UTXOs which falls within our range and minimizes waste.
+ *
+ * An additional optimization of this algorithm implemented here is a lookahead value which maintains the total value of the UTXO set
+ * of all unexplored UTXOs (i.e. UTXOs that have not yet been included or excluded). This allows us to cut a branch if the remaining
+ * amount is not sufficient to reach our target.
+ *
+ * SelectCoinsBnB Arguments:
+ * const std::vector<CInputCoin>& utxo_pool -> The set of UTXOs that we are choosing from. These UTXOs will be sorted in descending order
+ *                                             and the CInputCoins' values are their effective values.
+ * const CAmount& target_value -> This is the value that we want to select. It is the lower bound of the range.
+ * const CAmount& cost_of_change -> This is the cost of creating and spending a change output. This plus target_value is the upper bound
+ *                                  of the range.
+ * std::set<CInputCoin>& out_set -> This is an output parameter for the set of CInputCoins that have been selected.
+ * CAmount& value_ret -> This is an output parameter for the total value of the CInputCoins that were selected.
+ * CAmount& fee_ret -> This is an output parameter for the value of the transaction fees for the CInputCoins that were selected.
+ */
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount& fee_ret)
+{
+    out_set.clear();
+    value_ret = 0;
+
+    if (utxo_pool.empty()) {
+        return false;
+    }
+
+    int depth = 0;
+    int remaining_tries = 100000;
+    std::vector<std::pair<bool, bool>> selection; // First bool: select the utxo at this index; Second bool: traversing exclusion branch of this utxo
+    selection.assign(utxo_pool.size(), std::pair<bool, bool>(false, false));
+    bool done = false;
+    bool backtrack = false;
+
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+
+    // Calculate lookahead
+    CAmount lookahead = 0;
+    for (const CInputCoin& utxo : utxo_pool) {
+        lookahead += utxo.txout.nValue;
+    }
+
+    // Best solution
+    CAmount curr_waste = 0;
+    std::vector<std::pair<bool, bool>> best_selection;
+    CAmount best_waste = MAX_MONEY;
+
+    // Depth First search loop for choosing the UTXOs
+    while (!done)
+    {
+        if (remaining_tries <= 0) { // Too many tries, exit
+            break;
+        } else if (value_ret > target_value + cost_of_change) { // Selected value is out of range, go back and try other branch
+            backtrack = true;
+        } else if (curr_waste > best_waste) { // Don't select things which we know will be more wasteful
+            backtrack = true;
```

Comment:
Do we have a test that checks whether this behaves correctly for fee rates below the longterm expected fee rate? (I seem to remember that we discussed that before, and we came to the conclusion that it is correct, but if we had a test I'd be more comfortable.)